archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,261 @@
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : 100 ≤ n):
+  (∃ (a b c : ℕ), a ∈ finset.Ico n (2 * n + 1) ∧ b ∈ finset.Ico n (2 * n + 1) ∧
+  c ∈ finset.Ico n (2 * n + 1) ∧ (a : ℕ) < b ∧ (b : ℕ) < c ∧ (∃ (k : ℕ), (a : ℕ) + b = k * k) ∧
+  (∃ (l : ℕ), (c : ℕ) + a = l * l) ∧ (∃ (m : ℕ), (b : ℕ) + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : 107 ≤ n,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    by_cases p : 1 < l,
+    { have h₁ : 4 * l ≤ 2 * l * l, { linarith },
+      have h₂ :  2 * l * l - 4 * l ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        zify [h₁],
+        split,
+        { linarith },
+        { linarith }},
+      have h₃ : 2 * l * l + 4 * l ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        split,
+        { linarith },
+        { linarith }},
+      have h₄ : 2 * l * l - 4 * l < 2 * l * l + 1,
+      { zify [h₁],
+        linarith },
+      have h₅ : 2 * l * l + 1 < 2 * l * l + 4 * l, { linarith },
+      have h₆ : 2 * l * l + 1 ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        split,
+        { rw finset.Ico.mem at h₂,
+          linarith },
+        { rw finset.Ico.mem at h₃,
+          linarith }},
+      use [2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l],
+      refine ⟨h₂,⟨h₆,⟨h₃,⟨h₄,⟨h₅,⟨_,⟨_,_⟩⟩⟩⟩⟩⟩⟩,
+      { use (2 * l - 1),
+        have h₇ : 1 ≤ 2 * l, { linarith },
+        zify [h₁, h₇],
+        ring_exp_eq },
+      { use (2 * l),
+        zify [h₁],
+        ring_exp_eq },
+      { use(2 * l + 1),
+        ring_exp_eq }},
+    { exfalso,
+      simp only [not_lt] at p,
+      interval_cases l,
+      all_goals {linarith } }},
+  --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+  --construction of a triplet {a, b, c}, which is constructed by setting l=9
+  --in the argument at the start of the file.
+  { have h₈ : 126 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+      have h₉ : 163 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+    have h₁₀ : 198 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+    use [126, 163, 198],
+    refine ⟨h₈,⟨h₉,⟨h₁₀,⟨_,⟨_,⟨⟨17, _⟩,⟨⟨18, _⟩, ⟨19, _⟩⟩⟩⟩⟩⟩⟩⟩,
+    all_goals { linarith } },
+end
+
+lemma finset.Ico.mem_compl_iff_nmem_subset (a n : ℕ) (A : finset ℕ)
+  (h : a ∈ finset.Ico n (2 * n + 1)) : a ∉ A ↔ a ∈ finset.Ico n (2 * n + 1) \ A :=
+begin
+  rw finset.mem_sdiff,
+  split,
+  { intro ha,
+    exact ⟨h, ha⟩ },
+  { intro ha,
+    exact ha.2 },
+end
+
+theorem IMO_2021_Q1 : ∀ (n : ℕ), 100 ≤ n → ∀ (A ⊆ finset.Ico n (2 * n + 1)),
+  (∃ (a b : A), a ≠ b ∧ ∃ (k : ℕ), (a : ℕ) + b = k * k) ∨
+  (∃ (a b : finset.Ico n (2 * n + 1) \ A), a ≠ b ∧ ∃(k : ℕ), (a : ℕ) + b = k * k) :=
```

Comment:
Do not use subtypes here (you wrote `a : A` which means "`a` is a term of `A` coerced to a type", while you probably meant `a ∈ A` which means "`a` is an element of `A`".)
```suggestion
  (∃ a b ∈ A, a ≠ b ∧ ∃ (k : ℕ), a + b = k * k) ∨
  (∃ a b ∈ finset.Ico n (2 * n + 1) \ A, a ≠ b ∧ ∃ (k : ℕ), a + b = k * k) :=
```
This will require more changes downstream but it should make the proof easier.